### PR TITLE
Add validation tests for Groovy AST transformations

### PR DESCRIPTION
**Pull request by sclassen**
Add test for the following groovy AST annotations:
- `@AutoExternalize`
- `@IndexProperty`
- `@Bindable`
- `@Vetoable`

Tests are currently failing because of missing features in groovy
Thanks @vad1mo for the cool @Hackergarten in Zurich

### CODE REVIEW

**Comment by marchof** [comment]
@sclassen Thanks for your contribution! Is there a corresponding feature request at the groovy project we can track?

**Comment by Vad1mo** [comment]
yes 

https://issues.apache.org/jira/browse/GROOVY-9050
https://issues.apache.org/jira/browse/GROOVY-9051
https://issues.apache.org/jira/browse/GROOVY-9052
https://issues.apache.org/jira/browse/GROOVY-9053

**Comment by aalmiray** [comment]
All issues reported to the Groovy project have been fixed targeting Groovy `3.0.0-beta-1` and `2.5.7`.

File: org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/targets/GroovyIndexPropertyClassTarget.groovy [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+ *
+ *******************************************************************************/
+package org.jacoco.core.test.validation.groovy.targets
+
+import groovy.io.FileType
+import groovy.transform.IndexedProperty
+
+/* This annotation generates the following
+ * - writeExternal()
+ # - readExternal()
```
[the remark this answers is not in the document]
Comment (sclassen) [reply org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/targets/GroovyIndexPropertyClassTarget.groovy]: fixed

File: org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/GroovyVetoableClassTest.java [hunk 2]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ *******************************************************************************/
+package org.jacoco.core.test.validation.groovy;
+
+import org.jacoco.core.test.validation.ValidationTestBase;
+import org.jacoco.core.test.validation.groovy.targets.GroovyVetoableClassTarget;
+import org.junit.Test;
+
+/**
+ * Test of <code>data class</code>es.
```
[the remark this answers is not in the document]
Comment (sclassen) [reply org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/GroovyVetoableClassTest.java]: fixed

File: org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/GroovyIndexPropertyClassTest.java [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ *******************************************************************************/
+package org.jacoco.core.test.validation.groovy;
+
+import org.jacoco.core.test.validation.ValidationTestBase;
+import org.jacoco.core.test.validation.groovy.targets.GroovyIndexPropertyClassTarget;
+import org.junit.Test;
+
+/**
+ * Test of <code>data class</code>es.
```
[the remark this answers is not in the document]
Comment (sclassen) [reply org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/GroovyIndexPropertyClassTest.java]: fixed

File: org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/GroovySortableClassTest.java [hunk 4]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ *******************************************************************************/
+package org.jacoco.core.test.validation.groovy;
+
+import org.jacoco.core.test.validation.ValidationTestBase;
+import org.jacoco.core.test.validation.groovy.targets.GroovySortableClassTarget;
+import org.junit.Test;
+
+/**
+ * Test of <code>data class</code>es.
```
[the remark this answers is not in the document]
Comment (sclassen) [reply org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/GroovySortableClassTest.java]: fixed

File: org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/GroovyBindableClassTest.java [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ *******************************************************************************/
+package org.jacoco.core.test.validation.groovy;
+
+import org.jacoco.core.test.validation.ValidationTestBase;
+import org.jacoco.core.test.validation.groovy.targets.GroovyBindableClassTarget;
+import org.junit.Test;
+
+/**
+ * Test of <code>data class</code>es.
```
[the remark this answers is not in the document]
Comment (sclassen) [reply org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/GroovyBindableClassTest.java]: fixed

File: org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/GroovyAutoExternalizeClassTest.java [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+ *
+ *******************************************************************************/
+package org.jacoco.core.test.validation.groovy;
+
+import org.jacoco.core.test.validation.ValidationTestBase;
+import org.jacoco.core.test.validation.groovy.targets.GroovyAutoExternalizeClassTarget;
+import org.junit.Test;
+
+/**
+ * Test of <code>data class</code>es.
```
[the remark this answers is not in the document]
Comment (sclassen) [reply org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/GroovyAutoExternalizeClassTest.java]: fixed

File: org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/targets/GroovyAutoExternalizeClassTarget.groovy [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+ * - writeExternal()
+ # - readExternal()
+ */
+
+@groovy.transform.AutoExternalize
+class GroovyAutoExternalizeClassTarget { // assertEmpty()
+
+    String first    // assertEmpty()
+    List favItems     // assertEmpty()
+    Date since    // assertEmpty()
```
[the remark this answers is not in the document]
Comment (sclassen) [reply org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/targets/GroovyAutoExternalizeClassTarget.groovy]: I would like to leave the code as it is. It is an exact copy from the example in the documentation:
http://docs.groovy-lang.org/latest/html/gapi/groovy/transform/AutoExternalize.html

File: org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/targets/GroovySortableClassTarget.groovy [hunk 8]
```diff
@@ -0,0 +1,33 @@
...
+
+/* This annotation generates the following
+ * - implement the Comparable interface
+ * - have a compareTo method based on the first, last and born properties (priority ordering will be according to the ordering of property definition, highest first, unless 'includes' is used; in which case, priority will be according to the order given in the includes list)
+ * - have three Comparator methods named comparatorByFirst, comparatorByLast and comparatorByBorn
+ * - sort by natural order by default, reversed natural order can be specified
+
+ */
+
+@groovy.transform.Sortable
```
[the remark this answers is not in the document]
Comment (sclassen) [reply org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/targets/GroovySortableClassTarget.groovy]: fixed

File: org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/targets/GroovyAutoExternalizeClassTarget.groovy [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ *******************************************************************************/
+package org.jacoco.core.test.validation.groovy.targets
+
+/* This annotation generates the following
+ * - writeExternal()
+ # - readExternal()
+ */
+
+@groovy.transform.AutoExternalize
```
[the remark this answers is not in the document]
Comment (sclassen) [reply org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/targets/GroovyAutoExternalizeClassTarget.groovy]: fixed

File: org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/targets/GroovyVetoableClassTarget.groovy [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+/* This annotation generates the following
+ * - getters and setters
+ * - 2 versions of addVetoablePropertyChangeListener()
+ * - 2 versions of removeVetoablePropertyChangeListener()
+ * - 2 versions of getVetoablePropertyChangeListeners()
+ * - fireVetoablePropertyChange()
+ */
+
+@groovy.beans.Vetoable
```
[the remark this answers is not in the document]
Comment (sclassen) [reply org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/targets/GroovyVetoableClassTarget.groovy]: fixed

File: org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/targets/GroovyBindableClassTarget.groovy [hunk 11]
```diff
@@ -0,0 +1,34 @@
...
+ * - 2 versions of getPropertyChangeListeners()
+ * - firePropertyChange()
+ */
+
+class GroovyBindableClassTarget { // assertEmpty()
+
+    @groovy.beans.Bindable
+    String firstName    // assertEmpty()
+
+    @groovy.beans.Bindable
```
[the remark this answers is not in the document]
Comment (sclassen) [reply org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/targets/GroovyBindableClassTarget.groovy]: fixed

**Comment by sclassen** [comment]
@Godin can you please re-review, thx

File: org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/targets/GroovyAutoExternalizeClassTarget.groovy (hunk 7 above)
[the remark this answers is not in the document]
Comment (Godin) [reply org.jacoco.core.test.validation.groovy/src/org/jacoco/core/test/validation/groovy/targets/GroovyAutoExternalizeClassTarget.groovy]: One day someone might change this documentation 😈 but ok, let's keep as it is

**Comment by Godin** [comment]
@sclassen thank you for your contribution! @Vad1mo and @aalmiray thank you for Hackergarten! ❤️

**Pull request merged**